### PR TITLE
Code Review- Milo IDE V1.0

### PR DESCRIPTION
**Pull request by milobot**
## Features implemented:
- Basic JS blocks + code generation
- UI for workspace, code execution, XML
- ML Features
     - Loading in-built datasets (Iris, California Housing dataset)
     - Plotting - Histogram, Scatter plots
     - Images - from URL, Webcam, Squeezenet Classifier
     - Math - Deeplearn.js Array1D, Scalar Arithmetic, Array from Text
     - Stochastic Gradient Descent optimizer
     - Training SGD optimized model with data (using deeplearn.js)
     - Uploading custom datasets using CSVs
     - Bug fixes

### CODE REVIEW

File: views/dialogs.pug [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  .modal-dialog.modal-sm
+    .modal-content
+      .modal-header
+        h4#alertHeader.modal-title
+      .modal-body
+        p#alertBody(style="margin-left: 2px")
+      .modal-footer
+        button.btn.btn-default(type='button', data-dismiss='modal') Ok
+
+#confirmModal.modal.fade(tabindex='-1', role='dialog', aria-labelledby='mySmallModalLabel', aria-hidden='true')
```
Comment (arjun-rao) [inline views/dialogs.pug]: Remove aria-labelledby

File: views/editor.pug [hunk 2]
```diff
@@ -1,56 +1,73 @@
...
-      pre#source_javascript.code.
-
-      br
-      hr
-      |     Console Output:
-      #console_javascript
-    textarea#content_xml.content(wrap="off")
-    include toolbox.html
\ No newline at end of file
+	include head.pug
```
Comment (arjun-rao) [inline views/editor.pug]: Indent with 2 spaces

File: views/head.pug [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+	script(src="scripts/blockly.js")
+	script(src="scripts/blocks.js")
+	script(src="scripts/javascript.js")
+	script(src="scripts/dlearn.js")
+	script(src="scripts/custom/functions.js")
+	script(src="scripts/deeplearn.js")
+	script(src="scripts/plotly.js")
+	script(src="scripts/custom/codegen.js")
+	script(src="scripts/custom/blocks.js")
+	script(src="https://cdnjs.cloudflare.com/ajax/libs/jexcel/1.3.4/js/jquery.jexcel.js")
```
Comment (arjun-rao) [inline views/head.pug]: Include the script for better local development, instead of cdn

File: views/toolbox.html [hunk 4]
```diff
@@ -301,33 +302,99 @@
...
+        <block type="text_log">
+            <value name="TEXT">
+                <block type="dl_get_scalar">
+                  <value name="NUM">
+                  </value>
+                </block>
+            </value>
+          </block>
+      </category>
+      <category name="Miscellaneous" colour="%{BKY_ML_HUE}">
```
Comment (arjun-rao) [inline views/toolbox.html]: Remove unused blocks

File: public/scripts/code.js [hunk 5]
```diff
@@ -432,42 +431,49 @@ Code.initLanguage = function() {
...
+			throw MSG['timeout'];
+		}
+	};
+	var code = Blockly.JavaScript.workspaceToCode(Code.workspace);
+	Blockly.JavaScript.INFINITE_LOOP_TRAP = null;
+	try {
+		if(!window.navigator.onLine){
+			internetConnection.showOfflineAlert();
+		}
+			var setup =  DeepLearn.setup;
```
[the remark this answers is not in the document]
Comment (arjun-rao) [reply public/scripts/code.js]: Fixed

File: public/scripts/custom/codegen.js [hunk 6]
```diff
@@ -6,11 +7,232 @@ Blockly.JavaScript.constant = function(a) {
...
+            var row = [i,i*2];
+            Dataset.exceldata.push(row);
+        }
+        console.log(Dataset.exceldata);
+        var code = 'var inputX = []; \n var inputY = []; \n for(var i=0;i<1000;i++) \n { const arrX = dl.Array1D.new([i]); const arrY = dl.Array1D.new([i*2]); \n inputX.push(arrX);\n inputY.push(arrY); \n }';
+        var code1 = '\n var noAttributes = 1; \n ';
+        return code+code1;
+  };
+
+// Blockly.JavaScript['linearregression'] = function(block){
```
[the remark this answers is not in the document]
Comment (ayushbihani) [reply public/scripts/custom/codegen.js]: fixed

File: public/scripts/custom/helpers.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
+/**
+ * Namespace for internet connections functions,variables
+ */
+var internetConnection = {};
+internetConnection.isOnline = window.navigator.onLine;
```
[the remark this answers is not in the document]
Comment (ayushbihani) [reply public/scripts/custom/helpers.js]: Fixed

File: public/scripts/custom/helpers.js [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+internetConnection.showOfflineAlert = function(){
+    internetConnection.isOnline = false;
+    $('#alertHeader').html("Internet Disconnected");
```
[the remark this answers is not in the document]
Comment (ayushbihani) [reply public/scripts/custom/helpers.js]: Fixed

File: public/scripts/custom/helpers.js [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    $('#alertBody').html("Looks like you arent connected to the internet.<br>Some features may not work as expected!");
+    $('#alertModalTrigger').click();
+
+};
+
+internetConnection.showOnlineAlert = function(){
+    internetConnection.isOnline = true;
+    $('#alertHeader').html("Internet Connection Restored");
```
[the remark this answers is not in the document]
Comment (ayushbihani) [reply public/scripts/custom/helpers.js]: Fixed

File: views/toolbox.html (hunk 4 above)
Comment (ayushbihani) [reply views/toolbox.html]: Fixed
